Non-systemd udev path + adjusted unit test

red15 opened the pull request:
### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

Now only probing 2 paths at most.

This is a rewrite with lessons learned by writing #5180 .
Closes #3900

srebhan reviewed (approved):
While I would rather like to see the initialization of `ic.udevDataPath` in a `Init()` function, I can live with what you have now.

srebhan commented:
@red15: One thing, can you please rebase to run the CI checks again!? Thanks!

red15 commented:
> While I would rather like to see the initialization of `ic.udevDataPath` in a `Init()` function, I can live with what you have now.

Didn't know there was the ability have `Init()` function in a module like this, I'm not very familiar with the project as a whole.

> One thing, can you please rebase to run the CI checks again!? Thanks!

Done, although now the macOS CI is failing for some reason ?

srebhan commented:
Yeah, the example missing it but basically you can use
```
func (s *DiskIO) Init() error {
    ...
    return nil
}
```
The function is called after the config has been initialized, so it is a nice way to do the one-time things at construction...

Do you see a chance to address the `Init()` comment?

red15 commented:
> Do you see a chance to address the `Init()` comment?

Unlikely to happen soon, I'll need to setup my build environment again and I'm worried that the test scenario will make this more complicated as I'd have to make sure the test mocks are surviving the `Init()` phase.

srebhan commented:
I think we can live with the missing `Init()` and fix it later. Patches welcome!

helenosheaa commented:
@red15 can you please rebase this against master and temove telegraf from your CircleCI account as it's not running the mac test

red15 commented:
> @red15 can you please rebase this against master and temove telegraf from your CircleCI account as it's not running the mac test

Rebased on the master, I removed the CircleCI project on my side.

srebhan commented on plugins/inputs/diskio/diskio_linux.go:
```diff
@@ -37,9 +35,22 @@ func (s *DiskIO) diskInfo(devName string) (map[string]string, error) {
+		f, err := os.Open(udevDataPath)
+		defer f.Close()
+		if err != nil {
+			udevDataPath = fmt.Sprintf("/dev/.udev/db/block:%s", devName)
```
Should we also check if we can open that path and error out otherwise?

red15 replied:
Oops, looks like I resolved this without actually posting the response:

The original path presence check is still there, now on line 63 (not shown in the default diff on github)

srebhan replied:
Could you please move this test up to this location!? It's hard for a reviewer to jump through the code like this... :-)

red15 replied:
I considered keeping the code changes minimal more important as that would reduce the amount of risk in changing code paths, basically the whole idea of this patch was to assume the new udev would be present but gracefully handle the fallback to old /dev/.udev/db/block format which was unhandled before.

srebhan commented:
@helenosheaa any idea why the macdeps test is not executed?

helenosheaa commented:
@srebhan it's connected to his own telegraf CircleCi project so it's not executing.

@red15 it still seems to be linking to your CircleCi rather than ours, any ideas?

red15 commented:
> @red15 it still seems to be linking to your CircleCi rather than ours, any ideas?

This is because it's on my fork, (i'm owner) perhaps it would be easier to create a branch on the actual influxdata company/organization repository (i'm not able to create branches there), merge my change into that one then merge it to the  `influxdata/telegraf` `master` branch.

ssoroka commented:
@red15 it being on a fork shouldn't matter as a lot of people do that. Is it possible CircleCI created a Telegraf project for you and you need to delete it?

red15 commented:
As far as I can see I did already cancel the builds 3days ago, this is the last check worfklow that ran:
https://app.circleci.com/pipelines/github/red15/telegraf/328/workflows/740788ad-595e-49e1-8522-5d0c2ca7ad5e

I can't really see how to delete the telegraf project attached to my CircleCI account? If i look at my dashboard there are no followed projects.
